internal/graph/cached_resolver_test.go

```diff
@@ -515,3 +513,214 @@ func TestCachedCheckDatastoreQueryCount(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, uint32(1), res.GetResolutionMetadata().DatastoreQueryCount)
 }
+
+func TestCheckCacheKeyDoNotOverlap(t *testing.T) {
+	storeID := ulid.Make().String()
+	modelID := ulid.Make().String()
+
+	key1, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:x", "viewer", "user:jon"),
+	})
+	require.NoError(t, err)
+
+	key2, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:x", "viewer", "user:jon"),
+		ContextualTuples: []*openfgav1.TupleKey{
+			tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+		},
+	})
+	require.NoError(t, err)
+
+	contextStruct, err := structpb.NewStruct(map[string]interface{}{
+		"key1": true,
+	})
+	require.NoError(t, err)
+
+	key3, err := checkRequestCacheKey(&ResolveCheckRequest{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tuple.NewTupleKey("document:x", "viewer", "user:jon"),
+		Context:              contextStruct,
+	})
+	require.NoError(t, err)
+
+	// two Check request cache keys should not overlap if contextual tuples are
+	// provided in one and not the other and/or if context is provided in one
+	// and not the other
+	require.NotEqual(t, key1, key2)
+	require.NotEqual(t, key2, key3)
+	require.NotEqual(t, key1, key3)
+}
+
+func TestCheckCacheKeyWithContextualTuples(t *testing.T) {
+	storeID := ulid.Make().String()
+	modelID := ulid.Make().String()
+
+	tuples1 := []*openfgav1.TupleKey{
+		tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+		tuple.NewTupleKey("document:2", "viewer", "user:jon"),
+	}
+
+	tuples2 := []*openfgav1.TupleKey{
+		tuple.NewTupleKey("document:2", "viewer", "user:jon"),
+		tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+	}
```

Comment:
Is this `tuples1` but with the order inverted? So, test that the order of the ctx tuples don't impact the cache key?

Reply:
`[]*openfgav1.TupleKey` doesn't implement https://pkg.go.dev/sort#Interface, so it's not a valid target. If we were using primitive data types I could do that easily, but to do this I'd have to write more code than we already have here.

I renamed the test though.